Initial draft of upgrade guide for kubeadm clusters.

pipejakob opened the pull request:
This is a work-in-progress guide for upgrading kubeadm clusters from 1.6.x to 1.7.x.

Based on SIG feedback, I'm aiming for an initial MVP to get committed so that others can help iterate on it.

Fixes https://github.com/kubernetes/kubeadm/issues/278

@kubernetes/sig-cluster-lifecycle-pr-reviews

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/3999)
<!-- Reviewable:end -->

lukemarsden commented:
Thanks! Needs a link from the kubeadm guide to the upgrade guide.

timothysc commented on docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md:
```diff
@@ -0,0 +1,56 @@
+other 1.6.x releases. Upgrades are not supported before the 1.6.0 release (when
+kubeadm became Beta).
+
+It is a work-in-progress (not intended for merging yet) and targeted at other
```
I put the do-not-merge label on so you can remove this, plus the WIP in title also denotes.

timothysc commented on docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md:
```diff
@@ -0,0 +1,56 @@
+   a. On Debian, this would be:
+
+       sudo apt-get update
+       sudo apt-get install kubelet kubeadm kubelet kubernetes-cni
```
This should not be needed if it's an upgrade.

luxas replied:
Yeah, `sudo apt-get update && sudo apt-get upgrade` is sufficent

timothysc commented on docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md:
```diff
@@ -0,0 +1,56 @@
+   b. On CentOS/Fedora, this would be:
+
+       sudo yum update
+       sudo yum install kubelet kubeadm kubelet kubernetes-cni
```
same comment as the last one.

timothysc commented on docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md:
```diff
@@ -0,0 +1,56 @@
+
+       kubeadm init --skip-preflight-checks --kubernetes-version v1.7.0
+
+   For pre-release testing, this would be:
```
Nix from doc, I don't want this publicly visible honestly.

timothysc commented on docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md:
```diff
@@ -0,0 +1,56 @@
+
+4. Upgrade CNI provider.
+
+   Your CNI provider might have its own upgrade instructions to follow now.
```
Links to the major providers?

luxas replied:
There is an addons page, we might be able to link to that one...

timothysc commented on docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md:
```diff
@@ -0,0 +1,56 @@
+
+   Your CNI provider might have its own upgrade instructions to follow now.
+
+TODO: Rollback instructions in case anything goes wrong (using the backed up
```
cp -r backup /etc/kubernetes
yum downgrade x,y,z

luxas commented on docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md:
```diff
@@ -0,0 +1,56 @@
+---
+assignees:
+- pipejakob
+title: Upgrading kubeadm clusters to 1.7
```
from v1.6 to v1.7?

timothysc commented on docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md:
```diff
@@ -0,0 +1,56 @@
+
+This guide is for upgrading kubeadm clusters from version 1.6.x to 1.7.x, or
+other 1.6.x releases. Upgrades are not supported before the 1.6.0 release (when
+kubeadm became Beta).
```
@luxas going to need to add the warning re: https://github.com/kubernetes/kubernetes/pull/47081#discussion_r120518310

pipejakob commented:
I've addressed most of the feedback given, so please give this another look.

I still intend to add explicit rollback instructions, and I'll also host developer builds of debs for the most recent 1.7.0 beta release so others can actually test the instructions and file bugs for any problems found.

pipejakob commented:
I've built debs for `1.7.0-beta.2` for anyone willing to test these instructions, which are hosted in GCS:

    gs://beacham-public/kubernetes/1.7.0-beta.2/debs

There are subfolders for the different architectures, so to grab all of the debs for, say, amd64, you can run:

    mkdir kubernetes-debs
    gsutil rsync gs://beacham-public/kubernetes/1.7.0-beta.2/debs/amd64 kubernetes-debs

When following the instructions, instead of running `sudo apt-get update && sudo apt-get upgrade` you would instead run `sudo dpkg -i kubernetes-debs/*`.

If anyone wants to test the instructions using rpms, please let me know and I'll try to build those as well.

pipejakob commented:
I've gone ahead and built rpms for testing as well:

    gs://beacham-public/kubernetes/1.7.0-beta.2/rpms

timothysc commented on docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md:
```diff
@@ -0,0 +1,56 @@
+   a. On Debian, this can be accomplished with:
+
+       sudo apt-get update
+       sudo apt-get upgrade
```
I recommend just the update steps, it will keep around the legacy files in case there is one.

pipejakob replied:
I'm confused about the suggestion here. Are you suggesting to delete "sudo apt-get upgrade" and only leave the "sudo apt-get update" step? On Debian, that would only fetch the listing of new packages without upgrading any packages.

timothysc replied:
My debian is rusty, but on fedora/*EL you don't need the extra step so I'll comment below.

pipejakob replied:
Ah, got it. Will remove.

timothysc commented on docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md:
```diff
@@ -0,0 +1,56 @@
+
+       sudo yum update
+       sudo yum upgrade
+
```
You're missing a 
`sudo systemctl restart kubelet`

pipejakob replied:
Done!

pipejakob commented:
CC @lukemarsden

pipejakob commented:
We discussed this PR in this week's SIG Cluster Lifecycle meeting, since it is a hot topic for 1.7 burndown, and we were hoping it could get merged once a few people had tested the steps. If I can get ACKs from others that they've independently verified the process, will I need anything else other than a normal LGTM? I'm not sure how the "Needs Docs Review" label works.

Also, could someone please remove the `do-not-merge` label for me? I don't have permission. Thanks!

luxas commented on docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md:
```diff
@@ -0,0 +1,59 @@
+
+       sudo kubeadm init --skip-preflight-checks --kubernetes-version v1.7.0
+
+5. Upgrade CNI provider.
```
The admin has to delete the kube-proxy pods as well (so they get recreated)...
We aren't using rolling updates for that DS yet

pipejakob replied:
Good call-out. I'll add it to the doc and retest the steps.

pipejakob replied:
Done. However, the 1.7.0-beta.2 version doesn't seem to have the `kube-proxy` regression fix, so the pod ends up crash looping. I tested with kubeadm built from the HEAD of `release-1.7` and the steps worked fine.

The next 1.7 beta release is due out today, so I'll be able to build new debs/rpms for it and post them for testing. I'll also start building them for the HEAD of `release-1.7` so people can still test these instructions ASAP.

pipejakob commented:
> should we say something about how you may be able to restore things?

That was my original intention, but I was taking @lukemarsden's advice and trying to start minimal with the documentation so that we could keep adding to it once it was checked in. Now that it's been in review for so long, and we're so close to the release, I'll hedge a bit and open a second PR based on top of this commit to add the rollback instructions.

pipejakob commented:
Repeating this comment so it's not hidden in the collapsed, outdated diff:

I've added the step @luxas suggested for manually deleting `kube-proxy` so that it gets recreated during the `kubeadm init`. However, the 1.7.0-beta.2 version doesn't seem to have the kube-proxy regression fix, so the pod ends up crash looping. I tested with kubeadm built from the HEAD of release-1.7 and the steps worked fine.

The next 1.7 beta release is due out today, so I'll be able to build new debs/rpms for it and post them for testing. I'll also start building them for the HEAD of release-1.7 so people can still test these instructions ASAP.

timothysc commented on docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md:
```diff
@@ -0,0 +1,67 @@
+   b. On CentOS/Fedora, you would instead run:
+
+       sudo yum update
+       sudo yum upgrade
```
You don't need the extra upgrade step here, update works just fine and leaves breadcrumb config files in case you clobber something.

pipejakob replied:
Done.

pipejakob commented:
For anyone wanting to test the instructions using Debian on amd64, I've created new .debs here:

    gs://beacham-public/kubernetes/ci-1.7.0-beta.ccaf5152ae9f01/debs/amd64/

chenopis commented:
Ping me when there's a tech lgtm and this is ready to be merged. I'll make sure it gets into the 1.7 release.

pipejakob commented:
I've added steps to upgrade the OS packages on each node (and restart kubelet), along with headings to make it clear what steps to perform where.

Any other comments, @luxas or @timothysc?

luxas commented on docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md:
```diff
@@ -0,0 +1,87 @@
+
+### On the master
+
+1. Back up `/etc/kubernetes`.
```
Why should we have this if we don't tell how to restore?

pipejakob replied:
Good point. Ultimately, I'm hoping we can have backup, upgrade, and rollback instructions before the 1.7 release. I've already moved out the rollback instructions to a separate PR, so it probably makes sense to move the backup instructions there as well instead of leaving them dangling.

pipejakob commented:
@chenopis Tech LGTM is done by @luxas. Care to take a look?
